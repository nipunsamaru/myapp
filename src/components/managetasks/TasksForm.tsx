import {StyleSheet, View, Text, Alert} from 'react-native';
import React, {useState} from 'react';
import Input from './Input';
import Button from '../ui/Button';
import {getFormattedDate} from '../../util/date';
import {Colors} from '../../constants/style';

const TasksForm = ({submitButtonLabel, onCancel, onSubmit, defaultValues}) => {
  const [inputs, setInputs] = useState({
    taskName: {
      value: defaultValues ? defaultValues.taskName : '',
      isValid: true,
    },
    date: {
      value: defaultValues ? getFormattedDate(defaultValues.date) : '',
      isValid: true,
    },
    description: {
      value: defaultValues ? defaultValues.description : '',
      isValid: true,
    },
  });

  function inputChangedHandler(inputIdentifier, enteredValue) {
    setInputs(curInputs => {
      return {
        ...curInputs,
        [inputIdentifier]: {value: enteredValue, isValid: true},
      };
    });
  }
  function submitHandler() {
    const taskData = {
      taskName: inputs.taskName.value,
      date: new Date(inputs.date.value),
      description: inputs.description.value,
    };

    const taskIsValid = taskData.taskName.trim().length > 0;
    const dateIsValid = taskData.date.toString() !== 'Invalid Date';
    const descriptionIsValid = taskData.description.trim().length > 0;
    if (!taskIsValid || !dateIsValid || !descriptionIsValid) {
      //   Alert.alert('Invalid input', 'Please check your input values');
      setInputs(curInputs => {
        return {
          taskName: {value: curInputs.taskName.value, isValid: taskIsValid},
          date: {value: curInputs.date.value, isValid: dateIsValid},
          description: {
            value: curInputs.description.value,
            isValid: descriptionIsValid,
          },
        };
      });
      return;
    }

    onSubmit(taskData);
  }

  const formIsInvalid =
    !inputs.taskName.isValid ||
    !inputs.date.isValid ||
    !inputs.description.isValid;
  return (
    <View style={styles.form}>
      <Text style={styles.title}>Your Task</Text>
      <View style={styles.inputsRow}>
        <Input
          style={styles.rowInput}
          label="Task Name"
          invalid={!inputs.description.isValid}
          textInputConfig={{
            onChangeText: inputChangedHandler.bind(this, 'taskName'),
            value: inputs.taskName.value,
          }}
        />
        <Input
          style={styles.rowInput}
          label="Date"
          invalid={!inputs.description.isValid}
          textInputConfig={{
            placeholder: 'YYYY-MM-DD',
            maxLength: 10,
            onChangeText: inputChangedHandler.bind(this, 'date'),
            value: inputs.date.value,
          }}
        />
      </View>
      <Input
        label="Description"
        invalid={!inputs.description.isValid}
        textInputConfig={{
          multiline: true,
          // autoCapitalize: 'none'
          // autoCorrect: false // default is true
          onChangeText: inputChangedHandler.bind(this, 'description'),
          value: inputs.description.value,
        }}
      />
      {formIsInvalid && (
        <Text style={styles.errorText}>
          Invalid input values - please check your entered data!
        </Text>
      )}
      <View style={styles.buttons}>
        <Button style={styles.button} mode="flat" onPress={onCancel}>
          Cancel
        </Button>
        <Button style={styles.button} onPress={submitHandler}>
          {submitButtonLabel}
        </Button>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  form: {
    marginTop: 40,
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
    color: 'white',
    marginVertical: 24,
    textAlign: 'center',
  },
  inputsRow: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  rowInput: {
    flex: 1,
  },
  errorText: {
    textAlign: 'center',
    color: Colors.error500,
    margin: 8,
  },
  buttons: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  button: {
    minWidth: 120,
    marginHorizontal: 8,
  },
});

export default TasksForm;
