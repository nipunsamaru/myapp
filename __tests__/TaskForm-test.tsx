import React from 'react';
import renderer from 'react-test-renderer';
import TasksForm from '../src/components/managetasks/TasksForm';

describe('TasksForm component', () => {
  it('renders correctly', () => {
    const submitButtonLabel = 'Submit';
    const onCancel = jest.fn();
    const onSubmit = jest.fn();
    const defaultValues = {
      taskName: '',
      date: new Date(),
      description: '',
    };

    const tree = renderer
      .create(
        <TasksForm
          submitButtonLabel={submitButtonLabel}
          onCancel={onCancel}
          onSubmit={onSubmit}
          defaultValues={defaultValues}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  // Add more test cases as needed
});
