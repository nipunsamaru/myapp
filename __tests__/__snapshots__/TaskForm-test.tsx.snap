// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksForm component renders correctly 1`] = `
<View
  style={
    {
      "marginTop": 40,
    }
  }
>
  <Text
    style={
      {
        "color": "white",
        "fontSize": 24,
        "fontWeight": "bold",
        "marginVertical": 24,
        "textAlign": "center",
      }
    }
  >
    Your Task
  </Text>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        [
          {
            "marginHorizontal": 4,
            "marginVertical": 8,
          },
          {
            "flex": 1,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#c6affc",
              "fontSize": 12,
              "marginBottom": 4,
            },
            false,
          ]
        }
      >
        Task Name
      </Text>
      <TextInput
        onChangeText={[Function]}
        style={
          [
            {
              "backgroundColor": "#c6affc",
              "borderRadius": 6,
              "color": "#2d0689",
              "fontSize": 18,
              "padding": 6,
            },
          ]
        }
        value=""
      />
    </View>
    <View
      style={
        [
          {
            "marginHorizontal": 4,
            "marginVertical": 8,
          },
          {
            "flex": 1,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#c6affc",
              "fontSize": 12,
              "marginBottom": 4,
            },
            false,
          ]
        }
      >
        Date
      </Text>
      <TextInput
        maxLength={10}
        onChangeText={[Function]}
        placeholder="YYYY-MM-DD"
        style={
          [
            {
              "backgroundColor": "#c6affc",
              "borderRadius": 6,
              "color": "#2d0689",
              "fontSize": 18,
              "padding": 6,
            },
          ]
        }
        value="2024-03-07"
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "marginHorizontal": 4,
          "marginVertical": 8,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#c6affc",
            "fontSize": 12,
            "marginBottom": 4,
          },
          false,
        ]
      }
    >
      Description
    </Text>
    <TextInput
      multiline={true}
      onChangeText={[Function]}
      style={
        [
          {
            "backgroundColor": "#c6affc",
            "borderRadius": 6,
            "color": "#2d0689",
            "fontSize": 18,
            "padding": 6,
          },
          {
            "minHeight": 100,
            "textAlignVertical": "top",
          },
        ]
      }
      value=""
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "marginHorizontal": 8,
          "minWidth": 120,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
      >
        <View
          style={
            [
              {
                "backgroundColor": "#3e04c3",
                "borderRadius": 4,
                "padding": 8,
              },
              {
                "backgroundColor": "transparent",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "white",
                  "textAlign": "center",
                },
                {
                  "color": "#a281f0",
                },
              ]
            }
          >
            Cancel
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "marginHorizontal": 8,
          "minWidth": 120,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={false}
      >
        <View
          style={
            [
              {
                "backgroundColor": "#3e04c3",
                "borderRadius": 4,
                "padding": 8,
              },
              false,
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "white",
                  "textAlign": "center",
                },
                false,
              ]
            }
          >
            Submit
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
